rfcs/0017-intensional-store.md

```diff
@@ -0,0 +1,391 @@
+
+Other objective metadata that could be useful, such as late binding information, could be placed in this file too, perhaps as a TOML file that begins after the first double newline. This is to be determined.
+
+### Trust DB
+
+The Trust DB contains subjective metadata for `$out` -> `$cas` mappings. Each entry contains:
+
+- `$cas`
+- an array of `$out`
+- name (including version)
+- size
+- array of dependency `$cas`
+- optional:
+  - description
+  - build timestamp
+  - builder
+  - custom metadata, like one or more git repo commit hashes
+
+With this information, a user can quickly find `$cas` entries to install that match a name or description. `nix-build` can find a `$cas` by `$out`.
+
+For a given `$cas`, there can be many entries, one for each trusted source. This can be handled by having one SQLite DB per source (including localhost), and having an order of precedence.
+
+### Sharing the Nix Store
+
+Since the Nix Store (minus supporting directories) contains only self-validating paths, it can be shared "infinitely", only limited by:
+
+- disk space
+- network performance
+- confidence around hash collision attacks
+- confidence around writers corrupting paths without detection
+
+The installation step only involves moving a proposed path from `.prepare` to `.stage`, so no further communication is necessary with the Store daemon.
+
+For single-user installs, the Store can trivially be maintained by the Nix tools, and converting to multi-user is only a matter of changing the permissions.
+
+It would even be possible to use FUSE to automatically download any paths that are referenced in the Store, hanging the I/O request while it's being downloaded.
+
+### Store Daemon
+
+Optionally, a daemon can maintain the Store. In this case, it is recommended be the only user with write access. It performs installations, verifications and garbage collection, described below.
+
+### Preparing
+
+`$cas` entries are prepared by building them in a path that has the same total length as its final `$cas` entry. This means they can be built practically anywhere, in `/tmp`, in a home directory, in the `.prepare` directory in the Store, etc. It is recommended to make a part of the path a unique-per-build string.
+To make sure only the build's own references need rewriting, it is recommended to build using only `$cas` entries as dependencies, instead of relying on rewriting paths.
+
+After the build, the runtime dependencies file is generated in-memory (if there are dependencies), the `$cas` is calculated including the runtime dependencies buffer, any occurences of the build path are replaced with `/var/lib/nix/$cas`, and the `$cas.m` file is written (if there are dependencies).
+
+If there undetected build path references, they might cause the finished entry to work incorrectly, and they will cause `$cas` to differ on every build of `$out`. This must be handled on a case-by-case basis. Perhaps we'll need pluggable hash rewriters.
```

Comment:
We _absolutely_ need pluggable path rewriters, otherwise problems start at compressed manpages and only go worse from there!

Note that this is a _large_ increase of headache for some weird buildsystems compared with the status quo. Right now if a package can be installed by a non-root user by manually building it and dependencies, it has good chances to already have everything Nix needs. With CAS we need to make sure all the hidden references are rewritten (which, in the worst case, could lead to really annoying outcomes re: compression ratio and offsets, if you are rewriting an archive embedded at the end of an executable together with other archives…)

And we cannot check correctness unless we have bit perfect reproducibility (convergence of building+rewriting with the only difference being the build path is still not a guarantee, but pretty convincing evidence)

Reply:
Note that the post-processing can be part of the build, so "pluggable" could just mean "helper you call from the install phase".

OTOH, manpages are omnipresent and it would suck to have to alter all derivations. For them, I think the build env can just strip `$build/$out` altogether. When you're reading a manpage you know what executable it refers to, and having the long hashes in the manpages is just distracting, no?

So, a two-pronged approach with auto-detecting standard rewriters, and special-called-from-the-build rewriters?

Reply:
No, you cannot _just_ call the from the build, because you need to pass the final rewriter the computed content-addressed path to insert instead of the build path.

Documentation containing full path to an executable is indeed annoying; however, sometimes we end up with config files in store, and sometimes not even in the same output as the executable. So some paths are useful.

Then there is Java and JARs…

Reply:
This look like another field that could go in the metadata file, if `$cas.m` is some structured file like @kevincox suggested.
If a `$cas` needs a non standard rewriter it could note it in a specific field in the metadata and, if the nix daemon/setuid bin does not have support for that specific rewriter, it would report a the error without adding it to `/var/lib/nix`